execute/tokendata/usdc/http.go

```diff
@@ -0,0 +1,198 @@
+// oriented around caching/processing the attestation data instead of handling the API specifics.
+type httpClient struct {
+	apiURL     string
+	apiTimeout time.Duration
+	rate       *rate.Limiter
+	// coolDownUntil defines whether requests are blocked or not.
+	coolDownUntil time.Time
+	coolDownMu    *sync.RWMutex
+}
+
+//nolint:revive
+func NewHTTPClient(api string, apiInterval time.Duration, apiTimeout time.Duration) *httpClient {
+	return &httpClient{
+		apiURL:     api,
+		apiTimeout: apiTimeout,
+		rate:       rate.NewLimiter(rate.Every(apiInterval), 1),
+		coolDownMu: &sync.RWMutex{},
+	}
+}
+
+type attestationStatus string
+
+const (
+	attestationStatusSuccess attestationStatus = "complete"
+	attestationStatusPending attestationStatus = "pending_confirmations"
+)
+
+type httpResponse struct {
+	Status      attestationStatus `json:"status"`
+	Attestation string            `json:"attestation"`
+	Error       string            `json:"error"`
+}
+
+func (r httpResponse) validate() error {
+	if r.Error != "" {
+		return fmt.Errorf("attestation API error: %s", r.Error)
+	}
+	if r.Status == "" {
+		return fmt.Errorf("invalid attestation response")
+	}
+	if r.Status != attestationStatusSuccess {
+		return ErrNotReady
+	}
+	return nil
+}
+
+func (r httpResponse) attestationToBytes() ([]byte, error) {
+	attestationBytes, err := hex.DecodeString(strings.TrimPrefix(r.Attestation, "0x"))
+	if err != nil {
+		return nil, err
+	}
+	return attestationBytes, nil
+}
+
+func (h *httpClient) Get(ctx context.Context, messageHash [32]byte) ([]byte, int, error) {
+	var empty []byte
+	// Terminate immediately when rate limited
+	if h.inCoolDownPeriod() {
+		return empty, http.StatusTooManyRequests, ErrRateLimit
+	}
+
+	if h.rate != nil {
+		// Wait blocks until it the attestation API can be called or the
+		// context is Done.
+		if waitErr := h.rate.Wait(ctx); waitErr != nil {
+			return empty, http.StatusTooManyRequests, ErrRateLimit
+		}
+	}
+
+	timeoutCtx, cancel := context.WithTimeoutCause(ctx, h.apiTimeout, ErrTimeout)
+	defer cancel()
+
+	url := fmt.Sprintf("%s/%s/%s/0x%x", h.apiURL, apiVersion, attestationPath, messageHash)
+	return h.callAPI(timeoutCtx, url)
+}
+
+func (h *httpClient) callAPI(ctx context.Context, url string) ([]byte, int, error) {
+	var response []byte
+	// Use a timeout to guard against attestation API hanging, causing observation timeout and
+	// failing to make any progress.
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
+	if err != nil {
+		return response, http.StatusBadRequest, err
+	}
+	req.Header.Add("accept", "application/json")
+	res, err := http.DefaultClient.Do(req)
+	if err != nil {
+		if errors.Is(err, context.DeadlineExceeded) {
+			return response, http.StatusRequestTimeout, ErrTimeout
+		}
+		// On error, res is nil in most cases, do not read res.StatusCode, return BadRequest
+		return response, http.StatusBadRequest, err
```

Comment:
Should we just return `nil` here instead of `response` to ensure its never accessed? I think we had a bug around this in <V1.5

Reply:
`response` is not an http object, it's just a payload, no risk